src/wallet/wallet.cpp

```diff
@@ -3991,37 +3991,54 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
             return nullptr;
         }
         walletInstance->SetMaxVersion(nMaxVersion);
-    }
 
-    if (fFirstRun)
-    {
         // ensure this wallet.dat can only be opened by clients supporting HD with chain split and expects no default key
-        if (!gArgs.GetBoolArg("-usehd", true)) {
+        if (!gArgs.GetBoolArg("-usehd", true) && fFirstRun) {
             InitError(strprintf(_("Error creating %s: You can't create non-HD wallets with this version."), walletFile));
             return nullptr;
         }
-        walletInstance->SetMinVersion(FEATURE_NO_DEFAULT_KEY);
 
-        // generate a new master key
-        CPubKey masterPubKey = walletInstance->GenerateNewHDMasterKey();
-        if (!walletInstance->SetHDMasterKey(masterPubKey))
-            throw std::runtime_error(std::string(__func__) + ": Storing master key failed");
+        // Upgrade to HD if necessary
+        LOCK(walletInstance->cs_wallet);
```

Comment:
Perhaps you wanted this `LOCK` inside the scope that follows? As is, the lock is held during the call to `SetBestChain` below.

Reply:
Yes, that's what I wanted to do. Fixed.